README
Gitweb Lite

  (Beta release)
  gitweb.cgi clone. Serve git repository.
  Example: http://perlcodesample.com:10010

Features

  This is gitweb.cgi clone, and has the following features.

  - gitweb.cgi default features.
  - Written by Perl and Mojolicious web framework.
  - Prerequisite is Perl 5.8.7+.
  - Having HTTP server. You can run gitweblite,
    even if "apache" is not installed.
  - Multiple project homes.
  - Zero configuration.
  - Default encoding is UTF-8, you can change this.
  - don't work on Windows.

Operation

  Start
  
      ./gitweblite
  
    Application start in back ground. Port is 10010.
    You can access the following URL.
      
      http://localhost:10010
    
    If you change port, edit gitweblite.conf.
  
  Stop
  
      ./gitweblite --stop
    
    You can stop application by --stop option.

Description

  by default, /home and /git/pub directory is searched recursively.
  if xxx.git directory is found and it is git repository,
  the repository display on top page.

Config

  You can set config in gitweblite.conf.
  Config is written by JSON (// is comment line).
  Create your config file copied from gitweblite.conf.example
  
    cp gitweblite.conf.example gitweblite.conf

  These is available options.
  
    encoding
    
      Text encoding.
      
      Default:
      
        "encoding" : "UTF-8"

    git_bin
    
      git command (/usr/local/bin/git etc)
      
      Default:
      
        git command is automatically detected by PATH environment variable

    hypnotoad
    
      Server configuration.
    
      Default:
      
        "hypnotoad" : {
          "listen" : ["http://*:10010"]
        }
    
    logo_link
    
      Logo link URL
      
      Default:
      
      "logo_link" : "https://github.com/yuki-kimoto/gitweblite"
    
    prevent_xss
    
      Prevent XSS atack.
      
      Default:
      
        "prevent_xss" : 0
    
    search_dirs
        
      Git repository searching directories.
      
      Default:
      
        "search_dirs" : ["/git/pub", "/home"]
    
    search_max_depth
    
      Searching directory max depth.
      
      Default:
      
        "search_max_depth" : 10
    
    text_ext
    
      Text file extention. If "encoding" is not "UTF-8", generally text file
      is assumed as binary file. If "text_ext" is set, the file having the extention
      assumed as text file.
    
      Default:
      
        "text_exts" : ["txt"]

Logo and favicon

  You can use your logo and favicon.
  Set logo or favicon image file into the following name.
  
    user/public/git-logo.png
    user/public/git-favicon.png

COPYRIGHT & LICENSE

  Copyright 2009-2012 Yuki Kimoto, all rights reserved.

  This program is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.


Changes
0.05
  - added reverse proxy support
0.04
  - added encoding option
  - added text_exts option
  - added prevent_xss option
  - fixed many bugs.
0.03 (beta)
  - don't overwrite gitweblite.conf
  - config file become JSON.
  - URL become more restful.
  - Git command automatically detect
0.02 (beta)
  - fixed blob, blog_plain, commitdiff, commitdiff_plain encoding bug

0.01 (beta)
  - first beta release.


lib/Gitweblite.pm
use 5.008007;
package Gitweblite;

our $VERSION = '0.04';

use Mojo::Base 'Mojolicious';
use Gitweblite::Git;

has 'git';

sub startup {
  my $self = shift;
  
  # Config
  my $conf_file = $ENV{GITWEBLITE_CONFIG_FILE}
    || $self->home->rel_file('gitweblite.conf');
  $self->plugin('JSONConfigLoose', {file => $conf_file}) if -f $conf_file;
  my $conf = $self->config;
  $conf->{search_dirs} ||= ['/git/pub', '/home'];
  $conf->{search_max_depth} ||= 10;
  $conf->{logo_link} ||= "https://github.com/yuki-kimoto/gitweblite";
  $conf->{hypnotoad} ||= {listen => ["http://*:10010"]};
  $conf->{prevent_xss} ||= 0;
  $conf->{encoding} ||= 'UTF-8';
  $conf->{text_exts} ||= ['txt'];
  
  # Git
  my $git = Gitweblite::Git->new;
  my $git_bin = $conf->{git_bin} ? $conf->{git_bin} : $git->search_bin;
  die qq/Can't detect git command. set "git_bin" in gitweblite.conf/
    unless $git_bin;
  $git->bin($git_bin);
  $git->search_dirs($conf->{search_dirs});
  $git->search_max_depth($conf->{search_max_depth});
  $git->encoding($conf->{encoding});
  $git->text_exts($conf->{text_exts});
  $self->git($git);

  # Helper
  {
    # Remove top slash
    $self->helper('gitweblite_rel' => sub {
      my ($self, $path) = @_;
      
      $path =~ s/^\///;
      
      return $path;
    });
    
    # Get head commit id
    $self->helper('gitweblite_get_head_id' => sub {
      my ($self, $project) = @_;
      
      my $head_commit = $self->app->git->parse_commit($project, "HEAD");
      my $head_id = $head_commit->{id};
      
      return $head_id;
    });
  }
  
  # Added user public and templates path
  unshift @{$self->static->paths}, $self->home->rel_file('user/public');
  unshift @{$self->renderer->paths}, $self->home->rel_file('user/templates');
  
  # Reverse proxy support
  $ENV{MOJO_REVERSE_PROXY} = 1;
  $self->hook('before_dispatch' => sub {
    my $self = shift;
    
    if ( $self->req->headers->header('X-Forwarded-Host')) {
        my $prefix = shift @{$self->req->url->path->parts};
        push @{$self->req->url->base->path->parts}, $prefix;
    }
  });
    
  # Route
  my $r = $self->routes->route->to('main#');
    
  # Home
  $r->get('/')->to('#home');
  
  # Projects
  $r->get('/(*home)/projects')->to('#projects')->name('projects');
  
  # Project
  {
    my $r = $r->route('/(*project)', project => qr/.+?\.git/);
    
    # Summary
    $r->get('/summary')->to('#summary')->name('summary');
    
    # Short log
    $r->get('/shortlog/(*id)', {id => 'HEAD'})
      ->to('#log', short => 1)->name('shortlog');
    
    # Log
    $r->get('/log/(*id)', {id => 'HEAD'})->to('#log')->name('log');
    
    # Commit
    $r->get('/commit/(*id)')->to('#commit')->name('commit');
    
    # Commit diff
    $r->get('/commitdiff/(*diff)')->to('#commitdiff')->name('commitdiff');
    
    # Commit diff plain
    $r->get('/commitdiff-plain/(*diff)')
      ->to('#commitdiff', plain => 1)->name('commitdiff_plain');
    
    # Tags
    $r->get('/tags')->to('#tags')->name('tags');
    
    # Tag
    $r->get('/tag/(*id)')->to('#tag')->name('tag');
    
    # Heads
    $r->get('/heads')->to('#heads')->name('heads');
    
    # Tree
    $r->get('/tree/(*id_dir)', {id_dir => 'HEAD'})
      ->to('#tree')->name('tree');
    
    # Blob
    $r->get('/blob/(*id_file)')->to('#blob')->name('blob');
    
    # Blob plain
    $r->get('/blob-plain/(*id_file)')
      ->to('#blob', plain => 1)->name('blob_plain');
    
    # Blob diff
    $r->get('/blobdiff/(#diff)/(*file)')
      ->to('#blobdiff')->name('blobdiff');

    # Blob diff plain
    $r->get('/blobdiff-plain/(#diff)/(*file)')
      ->to('#blobdiff', plain => 1)->name('blobdiff_plain');
    
    # Snapshot
    $r->get('/snapshot/(:id)', {id => 'HEAD'})
      ->to('#snapshot')->name('snapshot');
  }
  
  # File cache
  $git->search_projects;
}

1;
